node/sync.go

```diff
@@ -272,6 +287,124 @@ func (d *Pegnetd) SyncBlock(ctx context.Context, tx *sql.Tx, height uint32) erro
 	return nil
 }
 
+// SnapshotPayouts moves the current shapshot to the "past", and updates the current snapshot. Then
+// it proceeds to do the snapshot staking payouts.
+func (d *Pegnetd) SnapshotPayouts(tx *sql.Tx, fLog *log.Entry, rates map[fat2.PTicker]uint64, height uint32, heightTimestamp time.Time) error {
+	// Snapshot
+	snapStart := time.Now()
+	err := d.Pegnet.SnapshotCurrent(tx)
+	if err != nil {
+		return err // Snapshot fails stop all progress and block syncing
+	}
+
+	// Payout snapshot
+	balances, err := d.Pegnet.SelectSnapshotBalances(tx)
+	if err != nil {
+		return err // Need to do staking payouts
+	}
+	staked := make(map[factom.FAAddress]*big.Int)
+	for _, bal := range balances {
+		// We want all balances in pUSD
+		total := new(big.Int)
+		for i := fat2.PTicker(1); i < fat2.PTickerMax; i++ {
+			if i == fat2.PTickerPEG {
+				continue // PEG does not count towards stake total
+			}
+			if bal.Balances[i] == 0 { // Ignore 0 balances
+				continue
+			}
+
+			// Convert from pXXX -> pUSD
+			c, err := conversions.Convert(int64(bal.Balances[i]), rates[i], rates[fat2.PTickerUSD])
+			if err != nil {
+				return err
+			}
+
+			// add c to running sum
+			total = total.Add(total, big.NewInt(c))
+		}
+
+		staked[*bal.Address] = total
+	}
+	// We need to mock a TXID for the staked payouts
+	txid := fmt.Sprintf("%064d", height)
+	// Sort the staked by highest PUSD
+	type StakedAmount struct {
+		Address factom.FAAddress
+		PUSD    uint64
+	}
+
+	var list []StakedAmount
+	for add, amt := range staked {
+		if !amt.IsUint64() {
+			return fmt.Errorf("%s has balance that is not uint64: %s", add, amt)
+		}
+
+		uAmt := amt.Uint64()
+		if uAmt <= 0 { // Apply a minimum required amount in pUSD
+			continue
+		}
+		// TODO: Check uint64 is safe
+		list = append(list, StakedAmount{Address: add, PUSD: uAmt})
+	}
+
+	if len(list) == 0 {
+		// Abort early since there is no one to pay out
+		fLog.WithFields(log.Fields{
+			"duration": time.Since(snapStart),
+			"eligible": len(list),
+		}).Info("balances snapshotted & staking not paid as there none eligible")
+	}
+
+	sort.Slice(list, func(i, j int) bool {
+		return list[i].PUSD < list[j].PUSD
+	})
+
+	// Calculate payouts
+	payoutindex := make(map[string]int)
+	addressMap := make(map[string]factom.FAAddress)
+
+	// 4.5K per block allowed
+	// as described in conversions
+	totalPayout := conversions.PerBlockAssetHolders * pegnet.SnapshotRate
```

Comment:
@sigrlami , `conversions.PerBlockAssetHolders` is invalid here because this branch is indicating v0.4.1.
go.mod should be updated with correct version.

Reply:
@StarNeit So, you probably confused my forked repo which my PR based of and have correct `pegnet` core version https://github.com/sigrlami/pegnetd/blob/pip-0012/go.mod with `pip-0012` developed by Steven back in April that still points to `v.0.4.1` which is irrelevant for PR. Version is correct and equal to the `staging` setup